File: src/sliding_window/LongestRepeatingCharReplacement.java
package sliding_window;

public class LongestRepeatingCharReplacement {

    public static void main(String[] args) {
        System.out.println();
        System.out.println(characterReplacement("XYYX", 2));
        System.out.println(characterReplacement("BAAABABB", 3));
        System.out.println(characterReplacement("AAABABB", 1));
        System.out.println(characterReplacement("ABBB", 2));
        System.out.println(characterReplacement("ABAB", 1));
    }

    public static int characterReplacement(String s, int k) {

        // Frequency array for 26 uppercase English letters
        int[] count = new int[26];
        int maxFreq = 0, res = 0, left = 0;

        for (int right = 0; right < s.length(); right++) {
            // Increment the count for the current character
            count[s.charAt(right) - 'A']++;
            maxFreq = Math.max(maxFreq, count[s.charAt(right) - 'A']);

            // Shrink the window if it exceeds the allowable replacements
            while ((right - left + 1) - maxFreq > k) {
                count[s.charAt(left) - 'A']--;
                left++;
            }

            // Update the result with the current window size
            res = Math.max(res, right - left + 1);
        }

        return res;
    }

}

File: src/sliding_window/LongestRepeatingCharReplacement.py
# class LongestRepeatingCharReplacement:
#
#   def characterReplacement(s: str, k: int) -> int:
#     # Frequency array for 26 uppercase English letters
#     count = [0] * 26
#     max_freq = 0
#     res = 0
#     left = 0
#
#     for right in range(len(s)):
#         # Increment the count for the current character
#         count[ord(s[right]) - ord('A')] += 1
#         max_freq = max(max_freq, count[ord(s[right]) - ord('A')])
#
#         # Shrink the window if it exceeds the allowable replacements
#         while (right - left + 1) - max_freq > k:
#             count[ord(s[left]) - ord('A')] -= 1
#             left += 1
#
#         # Update the result with the current window size
#         res = max(res, right - left + 1)
#     print(res)
#     return res
#
#   characterReplacement("ABABABBB",3)
#
#
#
